src/white.jl

```diff
@@ -0,0 +1,115 @@
+# the following conditions:
+#
+# The above copyright notice and this permission notice shall be
+# included in all copies or substantial portions of the Software.
+#
+# THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND,
+# EXPRESS OR IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF
+# MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE AND
+# NONINFRINGEMENT. IN NO EVENT SHALL THE AUTHORS OR COPYRIGHT HOLDERS BE
+# LIABLE FOR ANY CLAIM, DAMAGES OR OTHER LIABILITY, WHETHER IN AN ACTION
+# OF CONTRACT, TORT OR OTHERWISE, ARISING FROM, OUT OF OR IN CONNECTION
+# WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE SOFTWARE.
+
+export WhiteTest, BreuschPaganTest
+
+struct WhiteTest <: HypothesisTest
+    dof::Int                 #degrees of freedom in test
+    LM::Float64              #Lagrange Multiplier test statistic, distributed as Chisq(dof)
+    testtype::Symbol
+end
+
+"""
+    WhiteTest(X, e; testtype = :White)
+
+Compute White's (or Breusch-Pagan's) test for heteroskedasticity.
+
+`X` is a matrix of regressors and `e` is the vector of residuals from the original model.
+The keyword argument `testtype` is either `:linear` for the Breusch-Pagan/Koenker test,
+`:linear_and_squares` for White's test with linear and squared terms only (no cross-products), or
+`:White` (the default) for the full White's test (linear, squared and cross-product terms).
+`X` should include a constant and at least one more regressor. In some applications,
+`X` is a subset of the regressors in the original model, or just the fitted values.
+This saves degrees of freedom and may give a more powerful test. The LM test statistic
+is T*R2 where R2 is from the regression of `e^2` on the terms mentioned above.
+Under the null hypothesis it is distributed as `Chisq(dof)` where `dof` is the number
+of independent terms (not counting the constant), so the null is rejected when the
+test statistic is large enough.
+
+Implements: [`pvalue`](@ref)
+
+# References
+  * H. White, (1980): A heteroskedasticity-consistent covariance matrix estimator and a direct test for heteroskedasticity, Econometrica, 48, 817-838.
+  * T.S. Breusch & A.R. Pagan (1979), A simple test for heteroscedasticity and random coefficient variation, Econometrica, 47, 1287-1294
+  * R. Koenker (1981), A note on studentizing a test for heteroscedasticity, Journal of Econometrics, 17, 107-112
+
+# External links
+  * [White's test on Wikipedia](https://en.wikipedia.org/wiki/White_test)
+"""
+function WhiteTest(X::AbstractMatrix{<:Real}, e::AbstractVector{<:Real}; testtype = :White)
+
+    (n,K) = (size(X,1),size(X,2))    #nobs,nvars
+    n == length(e) || throw(DimensionMismatch("inputs must have the same length"))
+
+    K >= 2 || throw(ArgumentError("X must have >= 2 columns"))
+
+    intercept_cols = mapslices(col->all(diff(col).==0) && (first(col) != 0),X,dims=1)
+    any(intercept_cols) || throw(ArgumentError("one of the colums of X must be a non-zero constant"))
```

Comment:
This will avoid lots of allocations (I hope I got it right but you see the idea):
```suggestion
    intercept_col = any(eachcol(X)) do col
        first(col) != 0 && all(==(first(col)), col)
    end
    intercept_col || throw(ArgumentError("one of the colums of X must be a non-zero constant"))
```